client/rtc.go

```diff
@@ -86,62 +86,86 @@ func (c *Client) handleWSEventSignal(evData map[string]any) error {
 		}
 
-		c.log.Debug("received sdp offer", slog.Any("sdp", sdp))
-
-		if err := c.pc.SetRemoteDescription(webrtc.SessionDescription{
-			Type: webrtc.SDPTypeOffer,
-			SDP:  sdp,
-		}); err != nil {
-			return fmt.Errorf("failed to set remote description: %w", err)
-		}
-
-		answer, err := c.pc.CreateAnswer(nil)
-		if err != nil {
-			return fmt.Errorf("failed to create answer: %w", err)
-		}
-
-		if err := c.pc.SetLocalDescription(answer); err != nil {
-			return fmt.Errorf("failed to set local description: %w", err)
-		}
-
-		var sdpData bytes.Buffer
-		w := zlib.NewWriter(&sdpData)
-		if err := json.NewEncoder(w).Encode(answer); err != nil {
-			w.Close()
-			return fmt.Errorf("failed to encode answer: %w", err)
-		}
-		w.Close()
-		return c.SendWS(wsEventSDP, map[string]any{
-			"data": sdpData.Bytes(),
-		}, true)
+		return c.handleOffer(sdp)
 	case signalMsgAnswer:
 		sdp, ok := msg["sdp"].(string)
 		if !ok {
 			return fmt.Errorf("invalid SDP data received")
 		}
 
-		c.log.Debug("received sdp answer", slog.Any("sdp", sdp))
+		return c.handleAnswer(sdp)
+	default:
+		return fmt.Errorf("invalid signaling msg type %s", msgType)
+	}
 
-		if err := c.pc.SetRemoteDescription(webrtc.SessionDescription{
-			Type: webrtc.SDPTypeAnswer,
-			SDP:  sdp,
-		}); err != nil {
-			return fmt.Errorf("failed to set remote description: %w", err)
-		}
+	return nil
+}
 
-		for i := 0; i < len(c.iceCh); i++ {
-			c.log.Debug("adding queued remote candidate")
-			if err := c.pc.AddICECandidate(<-c.iceCh); err != nil {
-				return fmt.Errorf("failed to add remote candidate: %w", err)
-			}
+func (c *Client) handleAnswer(sdp string) error {
+	c.log.Debug("received sdp answer", slog.Any("sdp", sdp))
+
+	if err := c.pc.SetRemoteDescription(webrtc.SessionDescription{
+		Type: webrtc.SDPTypeAnswer,
+		SDP:  sdp,
+	}); err != nil {
+		return fmt.Errorf("failed to set remote description: %w", err)
+	}
+
+	for i := 0; i < len(c.iceCh); i++ {
+		c.log.Debug("adding queued remote candidate")
+		if err := c.pc.AddICECandidate(<-c.iceCh); err != nil {
+			return fmt.Errorf("failed to add remote candidate: %w", err)
 		}
-	default:
-		return fmt.Errorf("invalid signaling msg type %s", msgType)
 	}
 
 	return nil
 }
 
+func (c *Client) handleOffer(sdp string) error {
+	c.log.Debug("received sdp offer", slog.Any("sdp", sdp))
+
+	if err := c.pc.SetRemoteDescription(webrtc.SessionDescription{
+		Type: webrtc.SDPTypeOffer,
+		SDP:  sdp,
+	}); err != nil {
+		return fmt.Errorf("failed to set remote description: %w", err)
+	}
+
+	answer, err := c.pc.CreateAnswer(nil)
+	if err != nil {
+		return fmt.Errorf("failed to create answer: %w", err)
+	}
+
+	if err := c.pc.SetLocalDescription(answer); err != nil {
+		return fmt.Errorf("failed to set local description: %w", err)
+	}
+
+	if c.cfg.EnableDCSignaling && c.dc.ReadyState() == webrtc.DataChannelStateOpen {
+		c.log.Debug("sending answer through dc")
+		data, err := json.Marshal(answer)
```

Comment:
@cpoile Yeah. That's the upcoming PR that builds on top of this I mentioned the other day :p

Reply:
@cpoile I feel I've been a bit lazy with the above and should have shared a little more context for your personal knowledge if anything.

The primary reason we compress on WS is because we have that [8KB limit](https://github.com/mattermost/mattermost/blob/f3eee28f569ebdb58f5a79515bb74a6517ea6a0c/server/public/model/websocket_client.go#L21) on incoming messages into the MM server. On top of that, at the time of implementation, our SDPs were getting increasingly larger any time someone would screen share. That's not the case anymore since we now properly remove those tracks. Still, remote offers could get fairly large given audio tracks are never removed.

On DC, the theoretical limit is much higher, although browser implementations recommend not exceeding 16KB per message over the channel. That should be plenty, but compressing gives us also an advantage in case of low bandwidth situations (e.g. roaming mobile).